fvm/src/kernel.rs

```diff
@@ -0,0 +1,402 @@
+use crate::call_manager::KytheraCallManager;
+use crate::context::Override;
+use crate::machine::KytheraMachine;
+use cid::Cid;
+use fvm::call_manager::CallManager;
+use fvm::gas::{Gas, GasTimer, PriceList};
+use fvm::kernel::{
+    ActorOps, BlockId, BlockRegistry, BlockStat, CircSupplyOps, CryptoOps, DebugOps, EventOps,
+    GasOps, IpldBlockOps, LimiterOps, MessageOps, NetworkOps, RandomnessOps, SelfOps, SendOps,
+    SendResult,
+};
+use fvm::machine::Machine;
+use fvm::{DefaultKernel, Kernel};
+use fvm_shared::address::Address;
+use fvm_shared::clock::ChainEpoch;
+use fvm_shared::consensus::ConsensusFault;
+use fvm_shared::crypto::signature::{
+    SignatureType, SECP_PUB_LEN, SECP_SIG_LEN, SECP_SIG_MESSAGE_HASH_SIZE,
+};
+use fvm_shared::econ::TokenAmount;
+use fvm_shared::piece::PieceInfo;
+use fvm_shared::randomness::RANDOMNESS_LENGTH;
+use fvm_shared::sector::{
+    AggregateSealVerifyProofAndInfos, RegisteredSealProof, ReplicaUpdateInfo, SealVerifyInfo,
+    WindowPoStVerifyInfo,
+};
+use fvm_shared::sys::out::network::NetworkContext;
+use fvm_shared::sys::out::vm::MessageContext;
+use fvm_shared::sys::SendFlags;
+use fvm_shared::{ActorID, MethodNum};
+
+pub struct KytheraKernel<K = DefaultKernel<KytheraCallManager>> {
+    inner: K,
+}
+
+impl<M, C, K> Kernel for KytheraKernel<K>
+where
+    M: Machine,
+    C: CallManager<Machine = KytheraMachine<M>>,
+    K: Kernel<CallManager = KytheraCallManager<C>>,
+{
+    type CallManager = C;
+
+    fn into_inner(self) -> (Self::CallManager, BlockRegistry)
+    where
+        Self: Sized,
+    {
+        let (kythera_cm, br) = self.inner.into_inner();
+        (kythera_cm.0, br)
+    }
+
+    fn new(
+        mgr: Self::CallManager,
+        blocks: BlockRegistry,
+        caller: ActorID,
+        actor_id: ActorID,
+        method: MethodNum,
+        value_received: TokenAmount,
+        read_only: bool,
+    ) -> Self
+    where
+        Self: Sized,
+    {
+        Self {
+            inner: K::new(
+                KytheraCallManager(mgr),
```

Comment:
This file needs all these overrides mainly because of this right? I.e to be able to abstract the `CallManager`, can we make a comment stating that with a TODO to fix this upstream?